core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -81,8 +83,8 @@ abstract class MergingSnapshotProducer<ThisT> extends SnapshotProducer<ThisT> {
 
   // update data
   private final Map<PartitionSpec, List<DataFile>> newDataFilesBySpec = Maps.newHashMap();
-  private final CharSequenceSet newDataFilePaths = CharSequenceSet.empty();
-  private final CharSequenceSet newDeleteFilePaths = CharSequenceSet.empty();
+  private final DataFileSet newDataFiles = DataFileSet.create();
+  private final DeleteFileSet newDeleteFiles = DeleteFileSet.create();
```

Comment:
Do we need these extra collections? Can't we use sets in `newDataFilesBySpec` and `newDeleteFilesBySpec`?

Reply:
I'm handling this already in https://github.com/apache/iceberg/pull/11254/commits/4d42f180a4597d274f331dd5abe32fd840611f69. I just didn't want to introduce too many changes/refactorings as the PR is already quite large